MP OpenAPI - Adding tests for MP Rest Client integration

fabiobrz opened the pull request:
Integration tests to verify MP OpenAPI and MP Rest Client, see https://download.eclipse.org/microprofile/microprofile-open-api-1.1/microprofile-openapi-spec.html#_integration_with_other_microprofile_specifications

Jenkins job run:
https://eap-qe-jenkins.rhev-ci-vms.eng.rdu2.redhat.com/job/eap-7.x-microprofile-simple-face/8
but - as said - it is just for sanity checks, as `mp-open-api` profile is not taken into account in testing job, yet.
Tests are passing locally using the following feature branch:
https://github.com/pferraro/wildfly/tree/WFLY-12313
and the following TS branch:
https://github.com/fabiobrz/eap-microprofile-test-suite/tree/integration-mp-rest-client

Maven command:
```
mvn clean verify -Djboss.home=<path_to_built_feature_branch_wildfly_home> -Pmp-open-api -pl microprofile-open-api 
```


Please make sure your PR meets the following requirements:
- [x] Pull Request contains description for the changes
- [x] Pull Request does not include fixes for multiple issues / topics
- [x] Code is formatted, imports ordered, code compiles and tests are passing
- [x] Link to passing job is provided
- [x] Code is self-descriptive and/or documented
- [x] Description of the tests scenarios is included (see #46)

zhantemirov commented:
@fabiobrz 

First of all, I would like to note use of @tp* annotations - helps a lot in case a reader is not familiar with the thing that is being implemented.

Secondly, at first glance all looks good, but I will need to take a deeper look on Monday.

I didn't manage to run your tests locally, maybe I miss something.

Thanks.

fabiobrz commented:
Hi @zhantemirov thanks a lot for your work and comments. I added a Jenkins run (and related comments).

fabiobrz commented:
Hi @zhantemirov, I added a commit to FIX `ProgrammingModelTest`. Just moved a couple of methods inside an inner static class and added `@ServerSetup` annotation for it to be used. That's all so that we don't have to run another iteration for just that tiny fix.

honza-kasik commented on microprofile-open-api/src/test/java/org/jboss/eap/qe/microprofile/openapi/v11/MicroProfileOpenApi11Test.java:
```diff
@@ -144,8 +145,8 @@ public void testOpenApiDocumentForV11Annotations() {
         Assert.assertFalse("\"/districts/{code}\" \"patch\" property is empty", patchMethod.isEmpty());
-        Assert.assertTrue("\"/districts/{code}\" \"patch\" property should have exactly 8 keys",
-                patchMethod.keySet().size() == 8);
+        Assert.assertTrue("\"/districts/{code}\" \"patch\" property should have exactly 12 keys",
```
Please use `Assert.assertEquals` instead.

fabiobrz replied:
Fixing

honza-kasik commented on microprofile-open-api/src/main/java/org/jboss/eap/qe/microprofile/openapi/apps/routing/router/rest/LocalServiceRouterInfoResource.java:
```diff
@@ -0,0 +1,41 @@
+        try {
+            return Response.ok(InetAddress.getLocalHost().getHostName()).build();
+        } catch (UnknownHostException e) {
+            return Response.status(Response.Status.INTERNAL_SERVER_ERROR).build();
```
The `500` error code is not send when an exception is thrown on server so that's why are you returning it?

fabiobrz replied:
I get what you mean, probably this was definitely not necessary. Fixing, thanks.

honza-kasik commented on microprofile-open-api/src/test/java/org/jboss/eap/qe/microprofile/openapi/integration/restclient/RestClientIntegrationTest.java:
```diff
@@ -0,0 +1,160 @@
+package org.jboss.eap.qe.microprofile.openapi.integration.restclient;
+
+import static io.restassured.RestAssured.get;
+import static org.hamcrest.Matchers.*;
```
Please do not use wildcard imports.

fabiobrz replied:
Fixing

honza-kasik commented on microprofile-open-api/src/main/java/org/jboss/eap/qe/microprofile/openapi/apps/routing/router/services/DistrictServiceClient.java:
```diff
@@ -0,0 +1,53 @@
+package org.jboss.eap.qe.microprofile.openapi.apps.routing.router.services;
+
+import javax.ws.rs.*;
```
Please do not use wildcard imports.

fabiobrz replied:
Right, it must be my IDE that started doing this due to some kind of settings I've changed. Fixing.

fabiobrz commented on microprofile-open-api/src/main/java/org/jboss/eap/qe/microprofile/openapi/apps/routing/provider/rest/DistrictsResource.java:
```diff
@@ -49,13 +59,13 @@
     @Operation(summary = "Get all districts", description = "Retrieves and returns the available districts", operationId = "getAllDistricts")
     public Response getAllDistricts() {
-        return Response.ok().entity(districtService.getAll()).build();
+        return Response.ok().entity(districtService.getAll().stream().map(this::toEntity).collect(Collectors.toList())).build();
```
Sure, fixing

fabiobrz commented on microprofile-open-api/src/main/java/org/jboss/eap/qe/microprofile/openapi/apps/routing/router/rest/LocalServiceRouterInfoResource.java:
```diff
@@ -0,0 +1,41 @@
+import java.net.InetAddress;
+import java.net.UnknownHostException;
+
+import javax.ws.rs.*;
```
Right, fixing.

fabiobrz commented on microprofile-open-api/src/main/java/org/jboss/eap/qe/microprofile/openapi/apps/routing/router/rest/routed/RouterDistrictsResource.java:
```diff
@@ -0,0 +1,70 @@
+import java.net.URI;
+import java.net.URISyntaxException;
+
+import javax.ws.rs.*;
```
ok, fixing this one, too.

fabiobrz commented on microprofile-open-api/src/main/java/org/jboss/eap/qe/microprofile/openapi/apps/routing/router/model/DistrictObject.java:
```diff
@@ -0,0 +1,52 @@
+import java.io.Serializable;
+
+/**
+ * Represents a concrete entity
```
Thanks. Fixing class JavaDoc, copying getters and setters from `District` interface

fabiobrz commented on microprofile-open-api/src/main/java/org/jboss/eap/qe/microprofile/openapi/apps/routing/router/rest/routed/RouterDistrictsResource.java:
```diff
@@ -0,0 +1,70 @@
+
+    public RouterDistrictsResource() throws URISyntaxException {
+        serviceClient = RestClientBuilder.newBuilder()
+                .baseUri(new URI("http://localhost:8080/serviceProviderDeployment"))
```
Right, fixing.

fabiobrz commented on microprofile-open-api/src/main/java/org/jboss/eap/qe/microprofile/openapi/apps/routing/router/model/DistrictObject.java:
```diff
@@ -0,0 +1,52 @@
+
+    }
+
+    public DistrictObject(final String code, final String name, final Boolean obsolete) {
```
Sure.

fabiobrz commented:
> Thank you @fabiobrz! I reviewed the changes and I left some comments. Please also format the code so it can be reviewed without moving it horizontally (one long line -> several shorter ones) where possible.

Hi @honza-kasik, thanks for your review. I've pushed some changes to address your remarks. Please let me know if it's fine for you now.

rsvoboda commented:
I will merge once @honza-kasik approves
